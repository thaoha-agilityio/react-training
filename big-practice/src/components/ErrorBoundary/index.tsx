import { Component, ErrorInfo, ReactNode } from 'react';

interface Props {
  children: ReactNode;
  fallback: ReactNode;
}

interface State {
  hasError: boolean;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(): State {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error: Error, info: ErrorInfo): void {
    console.log(error);
    console.log(info);
  }

  render() {
    const { hasError } = this.state;
    const { children, fallback } = this.props;

    if (hasError) {
      // You can render any custom fallback UI
      return fallback;
    }

    return children;
  }
}

export default ErrorBoundary;
