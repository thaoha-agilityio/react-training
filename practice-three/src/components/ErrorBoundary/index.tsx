// libs
import { Component, ErrorInfo, ReactNode } from "react";

interface ErrorBoundaryProps {
  children: ReactNode;
}

interface ErrorBoundaryState {
  error: Error | null;
  errorInfo: ErrorInfo | null;
}

export class ErrorBoundary extends Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  public constructor(props: ErrorBoundaryProps | Readonly<ErrorBoundaryProps>) {
    super(props);
    this.state = { error: null, errorInfo: null };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    // Catch errors in any components below and re-render with error message
    this.setState({
      error,
      errorInfo,
    });
  }

  public render() {
    const { error, errorInfo } = this.state;
    const { children } = this.props;

    if (errorInfo) {
      return (
        <div>
          <h2>Oops !!</h2>
          <p>Something went wrong! Please try again later!</p>
          <p>{error && error.toString()}</p>
          <a href="/">Back to Home page</a>
        </div>
      );
    }

    return children;
  }
}
