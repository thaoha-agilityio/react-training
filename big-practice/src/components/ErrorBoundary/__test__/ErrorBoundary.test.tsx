import { render } from '@testing-library/react';

// Components
import ErrorBoundary from '..';

describe('Testing ErrorBoundary class', () => {
  it('Should render children when there is no error', () => {
    const { container, getByText } = render(
      <ErrorBoundary fallback={<div>Error occurred</div>}>
        <div>Content</div>
      </ErrorBoundary>,
    );
    expect(getByText('Content')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('Should render fallback when there is an error', () => {
    const ChildComponentWithError = ({ error }: { error: Error }) => {
      throw error;
    };
    const error = new Error('Test error');
    const { getByText } = render(
      <ErrorBoundary fallback={<div>Error occurred</div>}>
        <ChildComponentWithError error={error} />
      </ErrorBoundary>,
    );
    expect(getByText('Error occurred')).toBeInTheDocument();
  });
});
