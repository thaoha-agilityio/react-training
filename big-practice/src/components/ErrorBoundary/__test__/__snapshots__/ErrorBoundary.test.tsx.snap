// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ErrorBoundary class Should render children when there is no error 1`] = `
<div>
  <div>
    Content
  </div>
</div>
`;
